tools/autograd/gen_python_functions.py

```diff
@@ -427,9 +427,9 @@ def append_actuals_formals(actual, formal):
         env['actuals'] = actuals
 
         if has_tensor_options:
-            env['initialize_cuda'] = 'maybe_initialize_cuda(at::getType(options));'
+            env['initialize_cuda'] = 'maybe_initialize_cuda(options);'
         else:
-            env['initialize_cuda'] = 'maybe_initialize_cuda({});'.format(type_args[0]['name']) if type_args else ''
+            env['initialize_cuda'] = ''
```

Comment:
so if type_args what do you do?  Is it too late?  Do we need to turn it into an options?

Reply:
It's an inductive argument: if you have `type_args` and it's a CUDA type, you already had initialized it; no need to do it again :)

Reply:
I was asking something slightly different, sorry for not being clear.  How do you know you found all the places where we need to guard?  Why isn't there a single place we need to guard?  E.g. move at::getVariableType to torch::getVariableType and make sure that's the only way we can get an (uninitialized) VariableType and guard only that (don't take the specific suggestion too seriously, just trying to get the point across).

Reply:
> VariableType::getVariabletTypeFromBaseType

This doesn't work because VariableType lives in the Python agnostic part of Torch, whereas the initializer has to call into Python code.

I can introduce yet another function for getting types... but there will still be multiple places I would have to guard; for example, if someone calls `x.cuda()` I have no logical interposition point (this will go straight to the `cuda()` variable method) or they call `type.toBackend(kCUDA)` (this will go straight to the Type toBackend method).

I *suppose* what we could do instead is add a second CUDA initialization hook, so that when we require CUDA initialization, we call the first hook (which no-ops if it's not set), and then we call the second hook. That would definitively prevent any errors of this sort.